sympy/solvers/solveset.py

```diff
@@ -877,17 +877,35 @@ def _solve_radical(f, unradf, symbol, solveset_solver):
         result = Union(*[imageset(Lambda(y, g_y), f_y_sols)
                          for g_y in g_y_s])
 
-    if not isinstance(result, FiniteSet):
-        solution_set = result
```

Comment:
The previous code here skips all checking if the result isn't a FiniteSet but that misses cases like a Complement with a FiniteSet etc. The fix here makes the checking more comprehensive but is still incomplete. I covered all kinds of sets that emerged during the test suite.